src/digraph-cyclicity-hadwick-james.jl

```diff
@@ -0,0 +1,97 @@
+@doc_str """
+Find circuits (including self-loops) using the algorithm of Hadwick & James
+("Enumerating Circuits and Loops in Graphs with Self-Arcs and Multiple-Arcs", 2008)
+"""
+function simplecycles_hadwick_james end
+@traitfn function simplecycles_hadwick_james(g::::IsDirected)
+    nvg = nv(g)
+    B = [Int[] for i in 1:nvg]
+    blocked = [false for i in 1:nvg]
+    stack = Int[]
+    cycles = Vector{Vector{Int}}()
+    for v in vertices(g)
+        circuit_recursive!(g, v, v, blocked, B, stack, cycles)
+        resetblocked!(blocked)
+        resetB!(B)
+    end
+    return cycles
+end
+
+"""
+Reset B work structure.
+"""
+resetB!(B) = map!(empty!, B, B)
+
+"""
+Reset vector of "blocked" vertices.
+"""
+resetblocked!(blocked) = fill!(blocked, false)
+
+@doc_str """
+Find circuits starting from v1 recursively.
+"""
+function circuit_recursive! end
+@traitfn function circuit_recursive!(g::::IsDirected, v1::Int, v2::Int, blocked::Vector{Bool}, B::Vector{Vector{Int}}, stack::Vector{Int}, cycles::Vector{Vector{Int}})
```

Comment:
if `v1` and `v2` are meant to refer to vertex indices, they should be `T`, not `Int`. Also, with the above change, `blocked` is now a `BitArray`.

Also, you'll want to type-parameterize `B` with `T` (see above).